### PR TITLE
[5.1] Fix negative value validation and throw appropriate Exception on paginate and simplePaginate

### PR DESCRIPTION
**Pull request by unicodeveloper**
Solves this issue https://github.com/laravel/framework/issues/9960

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -268,6 +268,10 @@ public function lists($column, $key = null)
      */
     public function paginate($perPage = null, $columns = ['*'], $pageName = 'page', $page = null)
     {
+        if ($perPage < 0) {
+            throw new \Exception("Negative values can't be used to query a result set");
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: Please throw an invalid argument exception, and import it instead of using a slash.

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: Also, you need a fullstop at the end.

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: And you need to add a `@throws` docblock.

**Comment by busayo** [comment]
@GrahamCampbell  Fixed

**Comment by busayo** [comment]
@GrahamCampbell kindly check it out, any other stuff needed?

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 2]
```diff
@@ -265,9 +266,14 @@ public function lists($column, $key = null)
      * @param  string  $pageName
      * @param  int|null  $page
      * @return \Illuminate\Contracts\Pagination\LengthAwarePaginator
+     * @throws InvalidArgumentException
      */
     public function paginate($perPage = null, $columns = ['*'], $pageName = 'page', $page = null)
     {
+        if ($perPage < 0) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: `=<`

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 3]
```diff
@@ -288,9 +294,14 @@ public function paginate($perPage = null, $columns = ['*'], $pageName = 'page',
      * @param  array  $columns
      * @param  string  $pageName
      * @return \Illuminate\Contracts\Pagination\Paginator
+     * @throws InvalidArgumentException
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: cs

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 4]
```diff
@@ -265,9 +266,14 @@ public function lists($column, $key = null)
      * @param  string  $pageName
      * @param  int|null  $page
      * @return \Illuminate\Contracts\Pagination\LengthAwarePaginator
+     * @throws InvalidArgumentException
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: cs

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 5]
```diff
@@ -3,6 +3,7 @@
 namespace Illuminate\Database\Eloquent;
 
 use Closure;
+use InvalidArgumentException;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: cs

File: src/Illuminate/Database/Eloquent/Builder.php (hunk 2 above)
Comment (busayo) [inline src/Illuminate/Database/Eloquent/Builder.php]: U mean <= ?

**Comment by busayo** [comment]
@GrahamCampbell Fixed it using `php-cs-fixer fix Builder.php --level=psr2` .Don't know why styleCI is failing

**Comment by antonybudianto** [comment]
you need to add white space after exclamation mark on the if statement..

**Comment by GrahamCampbell** [comment]
> php-cs-fixer fix Builder.php --level=psr2

Because we're not just psr2.

**Comment by GrahamCampbell** [comment]
Don't bother running php-cs-fixer, just apply the diff from styleci.

**Comment by busayo** [comment]
@GrahamCampbell successfully fixed it! :smile:

**Comment by Sladewill** [comment]
If this is being considered in 5.1 then pages that are beyond the maximum amount of pages needs handling as well.

**Comment by busayo** [comment]
Should i go ahead and implement that? @GrahamCampbell @Sladewill

**Comment by Sladewill** [comment]
When I've tried that in the past its been rejected even though its the upper bounds of this, currently it doesn't work just like this so I would say it would be good but wait on Graham and Taylor.

**Comment by GrahamCampbell** [comment]
Stop changing all the test of the code.

**Comment by GrahamCampbell** [comment]
Stop changing all the rest of the code.

**Comment by GrahamCampbell** [comment]
Don't implement any extra checks. Please send a PR that just does what you said, and nothing else.

**Comment by busayo** [comment]
okay

**Comment by deefour** [comment]
This is broken logic. `$perPage` is now effectively required to be non-null and an int > 0. The fallback to the model's `getPerPage()` is never reached when `$perPage` is null.

```php
$perPage = $perPage ?: $this->model->getPerPage()
```

**Pull request closed without merging**